### PR TITLE
Refator test-documentation and add daily Travis CI

### PR DESCRIPTION
**Pull request by moretea**
- [x] Check when last CI job ran.
- [x] Add detection for all changes in docker images, or deployment scripts,  to trigger the correct platform tests.
- [x] Add platform test runner
- [x] Add documentation

### CODE REVIEW

**Comment by moretea** [comment]
Travis CI is going to schedule a CI job tomorrow at 1pm (see https://travis-ci.org/microservices-demo/microservices-demo/settings). Note that it starts the CI job from this branch. We could wait with merging until we're sure that it works on travis.

cc: @pidster

File: doc-tests/README.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+ 2. Hidden multiline annotation
+
+    &lt;!-- deploy-test-start PHASE [VALUE]*
+    CONTENT
+    --&gt;
+
+ 3. Visible multiline annotation
+
+    &lt;!-- deploy-test-start PHASE [VALUE]* --&gt;
```
Comment (moretea) [inline doc-tests/README.md]: These are not rendering properly. Will fix this.

File: doc-tests/README.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+The following syntaxes are supported to add snippets:
+
+ 1. Single line annotation, not visibile in Markdown output
+
+    &lt;!-- deploy-test PHASE [VALUE]* --&gt;
+
+ 2. Hidden multiline annotation
+
+    &lt;!-- deploy-test-start PHASE [VALUE]*
```
Comment (frankscholten) [inline doc-tests/README.md]: What about `deploy-test-hidden` to make it clear this step is hidden?

File: doc-tests/README.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
+# Testable Deployment Documentation
```
Comment (frankscholten) [inline doc-tests/README.md]: Let's use consistent terminology. How about Deployable Documentation? That way a Markdown file that adheres to this concept can be called a deploy doc.

File: docs/README.md [hunk 4]
```diff
@@ -1,6 +1,13 @@
+# Documentation
 This directory contains the deployment and internal documentation.
 
 Both are published on https://microservices-demo.github.io/microservices-demo/.
 
 The `local-server.sh` script starts a jekyll server to facilitate previews of the pages.
```
Comment (frankscholten) [inline docs/README.md:6]: Refactor to `local-jekyll.sh`

File: docs/deployment/ecs.md [hunk 5]
```diff
@@ -1,5 +1,7 @@
 ---
 layout: default
+executableDocumentation: true
+deploymentScriptDir: "aws-ecs"
 ---
```
Comment (frankscholten) [inline docs/deployment/ecs.md:5]: Same comment as above. I suggest to use the term deployable documentation.

File: doc-tests/README.md (hunk 2 above)
Comment (moretea) [reply doc-tests/README.md]: no, it's the html comment that has not ended yet.

two  online comments

```
<!-- one line comment 1 -->
CONTENT A
<!-- one line comment 2-->
```

vs 

```
<!-- several  
oneline
CONTENT B
comments
-->
```

Therefore  `CONTENTA` will be shown in the HTML, the and `CONTENTB` will not.

**Comment by moretea** [comment]
W.r.t. the term, I think it should somehow encompass that we're trying to make sure that the documentation is correct. I think "deployable documentation" looses that part.

Some alternatives:
- Verifiable deployment instructions
- Verified Documentation
- Verified and deployable Documentation
- Checked deployment steps
- Checked Docs

File: doc-tests/lib/travis.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
+require "json"
```
Comment (vlal-zz) [inline doc-tests/lib/travis.rb]: Should we use [this](https://github.com/travis-ci/travis.rb) library by travis to interact with travis?

File: doc-tests/bin/cronjob-test-job-spawner [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                        s
+                      else
+                        s = (Date.today - 30).to_time
+                        log(:info, "No cron job found! Checking since #{s.rfc2822}")
+                        s
+                      end
+
+
+# Find all executable documentation in $GIT_ROOT/doc/deployment.
+documented_deployments = ExecutableDeploymentDocumentation.find_all
```
Comment (vlal-zz) [inline doc-tests/bin/cronjob-test-job-spawner]: I can't find `ExecutableDeploymentDocumentation` defined anywhere. Also I think it should be s/doc/docs in the above comment.

File: doc-tests/lib/find_changes.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    .map { |part| part["spec"]["template"]["spec"]["containers"].map { |c| c["image"] } }.flatten\
+    .map { |name| if name.include?(":") then name; else name + ":latest"; end }.uniq
+end
+
+# Checks if any of the +image_list+ images on the docker hub have been changed since +since+.
+def any_image_changed_since?(image_list, since)
+  log(:info, "Checking if any of #{image_list.join(", ")} has changed since #{since.rfc2822}")
+
+  changed_images = image_list.select do |img|
+    log_system_or_die(:debug, "docker pull #{img}")
```
Comment (vlal-zz) [inline doc-tests/lib/find_changes.rb]: Another option is to hit the API directly. For example the payments url [here](https://hub.docker.com/v2/repositories/weaveworksdemos/payment/), which returns json with a `last_updated` field.

File: doc-tests/lib/travis.rb (hunk 6 above)
Comment (moretea) [reply doc-tests/lib/travis.rb]: That introduces a dependency. At the moment, I only depend on ruby's standard library. Also, I require the use of the 'requests' API endpoint, which is not yet implemented in the library.

File: doc-tests/lib/find_changes.rb (hunk 8 above)
Comment (moretea) [reply doc-tests/lib/find_changes.rb]: Perfect! This was my initial approach, but could not find that endpoint in the documentation. This definitely has my preference.

File: doc-tests/bin/cronjob-test-job-spawner (hunk 7 above)
Comment (moretea) [reply doc-tests/bin/cronjob-test-job-spawner]: Ah, typo yes.

File: doc-tests/lib/find_changes.rb (hunk 8 above)
Comment (moretea) [reply doc-tests/lib/find_changes.rb]: Ok, not going as fast as I had hoped.

Stuck at

```
[maarten@maarten-laptop:~]$ curl -v -I https://hub.docker.com/v2/repositories/library/mongo/manifests/3
*   Trying 52.71.230.17...
* TCP_NODELAY set
* Connected to hub.docker.com (52.71.230.17) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: OU=GT68254618; OU=See www.rapidssl.com/resources/cps (c)13; OU=Domain Control Validated - RapidSSL(R); CN=*.docker.com
*  start date: Oct 29 21:14:42 2014 GMT
*  expire date: Nov 13 14:19:33 2017 GMT
*  subjectAltName: host "hub.docker.com" matched cert's "*.docker.com"
*  issuer: C=US; O=GeoTrust Inc.; CN=RapidSSL SHA256 CA - G3
*  SSL certificate verify ok.
> HEAD /v2/repositories/library/mongo/manifests/3 HTTP/1.1
> Host: hub.docker.com
> User-Agent: curl/7.50.3
> Accept: */*
> 
< HTTP/1.1 404 NOT FOUND
HTTP/1.1 404 NOT FOUND
< Date: Fri, 04 Nov 2016 12:23:21 GMT
Date: Fri, 04 Nov 2016 12:23:21 GMT
< Content-Type: text/html; charset=utf-8
Content-Type: text/html; charset=utf-8
< X-Frame-Options: deny
X-Frame-Options: deny
< Strict-Transport-Security: max-age=31536000
Strict-Transport-Security: max-age=31536000
< Server: nginx
Server: nginx
< X-Content-Type-Options: nosniff
X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
X-XSS-Protection: 1; mode=block
< Strict-Transport-Security: max-age=31536000
Strict-Transport-Security: max-age=31536000

< 
* Curl_http_done: called premature == 0
* Connection #0 to host hub.docker.com left intact
```

Getting info on a specific tag is hard.

**Comment by moretea** [comment]
@frankscholten @vlal any remaining comments?

**Comment by vlal** [comment]
Nope. LGTM 🐐

**Comment by vlal-zz** [comment]
Curious it seems some of the changes are missing. `ExecutableDeploymentDocumentation` seems to be only defined in one place.

**Comment by pidster** [comment]
@vlal @moretea what's required before this can be merged?

**Comment by moretea** [comment]
@vlal AFAICS, there is no reference to `ExecutableDeploymentDocumentation` anywhere.

```
[maarten@maarten-laptop:~/cs/weave/microservices-demo/deploy-doc-tests]$ grep "ExecutableDeploymentDocumentation" . -R | wc -l
0

[maarten@maarten-laptop:~/cs/weave/microservices-demo/deploy-doc-tests]$ 
```

File: deploy-doc-tests/lib/deploy_doc_test.rb [hunk 9]
```diff
@@ -0,0 +1,44 @@
+require "yaml"
+class DeployDocTest < Struct.new(:name, :markdown_file, :deployment_directory, :config)
+  require_relative "deploy_doc_test/plan"
+
+  def self.find!(name)
+    deploy_doc_test = DeployDocTest.find_all.find { |deploy_doc_test| deploy_doc_test.name == name }
+    if deploy_doc_test.nil?
+      $stderr.puts "Executable deployment documentation for #{name} not found."
```
Comment (frankscholten) [inline deploy-doc-tests/lib/deploy_doc_test.rb]: DeployDoc for #{name} not found

**Review by frankscholten — APPROVED** [review]
LGTM

**Pull request merged**